src/git.rs

```diff
@@ -62,9 +62,41 @@ pub struct Crate {
     pub deps: Vec<Dependency>,
     pub cksum: String,
     pub features: HashMap<String, Vec<String>>,
+    /// This field contains features with new, extended syntax. Specifically,
+    /// namespaced features (`dep:`) and weak dependencies (`pkg?/feat`).
+    ///
+    /// It is only populated if a feature uses the new syntax. Cargo merges it
+    /// on top of the `features` field when reading the entries.
+    ///
+    /// This is separated from `features` because versions older than 1.19
+    /// will fail to load due to not being able to parse the new syntax, even
+    /// with a `Cargo.lock` file.
+    #[serde(skip_serializing_if = "Option::is_none")]
+    pub features2: Option<HashMap<String, Vec<String>>>,
     pub yanked: Option<bool>,
     #[serde(default)]
     pub links: Option<String>,
+    /// The schema version for this entry.
+    ///
+    /// If this is None, it defaults to version 1. Entries with unknown
+    /// versions are ignored by cargo starting with 1.51.
+    ///
+    /// Version `2` format adds the `features2` field.
+    ///
+    /// This provides a method to safely introduce changes to index entries
+    /// and allow older versions of cargo to ignore newer entries it doesn't
+    /// understand. This is honored as of 1.51, so unfortunately older
+    /// versions will ignore it, and potentially misinterpret version 2 and
+    /// newer entries.
+    ///
+    /// The intent is that versions older than 1.51 will work with a
+    /// pre-existing `Cargo.lock`, but they may not correctly process `cargo
+    /// update` or build a lock from scratch. In that case, cargo may
+    /// incorrectly select a new package that uses a new index format. A
```

Comment:
I'm not sure if there is a precedent for this (having editions apply beyond `rustc`), but I wonder if upon publish we should require the crate to be `edition = "2021"` if `features2` is used.

With this requirement we would ensure that if older versions of cargo do select a crate with features it doesn't understand, then the compilation would bail out early due to the new edition, rather than potentially completing successfully but with an unexpected selection of features.

Reply:
Hm, that's an interesting idea.  However, I'm not sure if it will help in enough situations to justify adding the restriction.  If the package required the feature to be set, it would fail during resolution before the package gets downloaded.  If somehow it makes it through resolution, I think at least versions older than 1.27 will fail to parse the manifest.  We are also planning to suggest only using this new syntax in a major version bump which should further limit the impact.  I would kinda prefer to avoid adding more restrictions if possible.  Do you feel strongly about trying to limit the confusion here?